### PR TITLE
[major] - Upgrade deps

### PR DESCRIPTION
**Pull request by 311chaos**
## Change Type

* [ ] Feature
* [x] Chore
* [ ] Bug Fix

## Change Level

* [x] major
* [ ] minor
* [ ] patch

## Further Information (screenshots, bug report links, etc)
* Major upgrade to eslint from 6 to 7. https://eslint.org/docs/user-guide/migrating-to-7.0.0#migrating-to-v700
* Eslint is currently at version 8 so we should look to update that in later PR. https://eslint.org/docs/8.0.0/user-guide/migrating-to-8.0.0
* Some of the rules from eslint-plugin-unicorn have been adjusted ([compare](https://github.com/sindresorhus/eslint-plugin-unicorn/compare/v24.0.0...v40.0.0)) so this may pick up issues that were previously not. We went from v24 to [v40](https://github.com/sindresorhus/eslint-plugin-unicorn/releases). In those versions a bunch of new rules have been added. We should consider some of these to our recommended package. I think we could start with: `empty-brace-spaces`, `prefer-date-now`, `prefer-default-parameters`, `no-new-array`, `consistent-destructuring`, `no-invalid-remove-event-listener`, `no-empty-file`.

 24.0.0

* [no-lonely-if](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/no-lonely-if.md)
* [empty-brace-spaces](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/empty-brace-spaces.md)
* [prefer-date-now](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/prefer-date-now.md)

25.0.0

* [prefer-array-some](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/master/docs/rules/prefer-array-some.md)
* [prefer-default-parameters](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/master/docs/rules/prefer-default-parameters.md)

26.0.0

* [no-new-array](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/master/docs/rules/no-new-array.md)
* [prefer-array-index-of](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/master/docs/rules/prefer-array-index-of.md)
* [prefer-regexp-test](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/master/docs/rules/prefer-regexp-test.md)
* [consistent-desctucturing](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/master/docs/rules/consistent-destructuring.md)

27.0.0

* [no-array-for-each](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/master/docs/rules/no-array-for-each.md)
* [no-array-push-push](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/master/docs/rules/no-array-push-push.md)
* [no-this-assignment](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/master/docs/rules/no-this-assignment.md)

29.0.0

* [no-static-only-class](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/no-static-only-class.md)
* [prefer-array-flat](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/prefer-array-flat.md)

30.0.0

* [prefer-switch](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/prefer-switch.md)

31.0.0

* [prefer-node-protocol](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/prefer-node-protocol.md)
* [prefer-module](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/prefer-module.md)

32.0.0

* [no-document-cookie](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/no-document-cookie.md)
* [numeric-separators-style](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/numeric-separators-style.md)
* [prefer-array-flat](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/prefer-array-flat.md)
* [prefer-array-flat-map](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/prefer-array-flat-map.md)

33.0.0

* [require-array-join-separator](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/require-array-join-separator.md)
* [require-number-to-fixed-digits-argument](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/require-number-to-fixed-digits-argument.md)
* [prefer-prototype-methods](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/prefer-prototype-methods.md)
* [prefer-object-has-own](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/prefer-object-has-own.md)

34.0.0

* [no-array-method-this-argument](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/no-array-method-this-argument.md)
* [require-post-message-target-origin](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/require-post-message-target-origin.md)
* [prefer-top-level-await](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/prefer-top-level-await.md)
* [prefer-at](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/prefer-at.md)

35.0.0

* [prefer-object-from-entries](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/prefer-object-from-entries.md)
* [no-useless-length-check](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/no-useless-length-check.md)
* [no-useless-spread](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/no-useless-spread.md)

36.0.0

* [no-useless-fallback-in-spread](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/no-useless-fallback-in-spread.md)
* [no-invalid-remove-event-listener](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/no-invalid-remove-event-listener.md)

37.0.0

* [template-indent](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/template-indent.md)

38.0.0

* [no-empty-file](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/no-empty-file.md)
* [prefer-export-from](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/prefer-export-from.md)

39.0.0

* [prefer-code-point](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/prefer-code-point.md)
* [no-await-expression-member](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/no-await-expression-member.md)

40.0.0

* [no-thenable](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/no-thenable.md)
* [no-useless-promise-resolve-reject](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/no-useless-promise-resolve-reject.md)
* [relative-url-style](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/relative-url-style.md)
* [prefer-json-parse-buffer](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/prefer-json-parse-buffer.md)

### CODE REVIEW

File: rules/recommended.js [hunk 1]
```diff
@@ -144,17 +144,18 @@ const baseRules = {
   'unicorn/filename-case': [2, { case: 'kebabCase' }],
   'unicorn/new-for-builtins': 2,
   'unicorn/no-abusive-eslint-disable': 2,
-  'unicorn/no-array-instanceof': 2,
+  'unicorn/no-document-cookie': 2,
```
Comment (311chaos) [inline rules/recommended.js:147]: new rule.

File: rules/recommended.js [hunk 2]
```diff
@@ -144,17 +144,18 @@ const baseRules = {
...
+  'unicorn/no-instanceof-array': 2,
```
Comment (311chaos) [inline rules/recommended.js:148]: this was renamed.

File: rules/recommended.js [hunk 3]
```diff
@@ -144,17 +144,18 @@ const baseRules = {
...
   'unicorn/no-useless-undefined': 2,
   'unicorn/no-zero-fractions': 2,
   'unicorn/prefer-array-find': 2,
   'unicorn/prefer-date-now': 2,
+  'unicorn/prefer-default-parameters': 2,
   'unicorn/prefer-includes': 2,
   'unicorn/prefer-optional-catch-binding': 2,
-  'unicorn/prefer-spread': 2,
-  'unicorn/prefer-starts-ends-with': 2,
+  'unicorn/prefer-string-starts-ends-with': 2,
```
Comment (311chaos) [inline rules/recommended.js:158]: rule was renamed.

File: rules/recommended.js [hunk 4]
```diff
@@ -144,17 +144,18 @@ const baseRules = {
...
-  'unicorn/no-array-instanceof': 2,
+  'unicorn/no-document-cookie': 2,
+  'unicorn/no-instanceof-array': 2,
   'unicorn/no-new-buffer': 2,
   'unicorn/no-unreadable-array-destructuring': 2,
   'unicorn/no-useless-undefined': 2,
   'unicorn/no-zero-fractions': 2,
   'unicorn/prefer-array-find': 2,
   'unicorn/prefer-date-now': 2,
+  'unicorn/prefer-default-parameters': 2,
```
Comment (311chaos) [inline rules/recommended.js:155]: new rule

File: rules/typescript.js [hunk 5]
```diff
@@ -62,7 +62,12 @@ const typescriptRules = {
   '@typescript-eslint/prefer-namespace-keyword': 0,
   '@typescript-eslint/prefer-string-starts-ends-with': 2,
   '@typescript-eslint/require-array-sort-compare': 2,
-  '@typescript-eslint/restrict-plus-operands': 2,
+  '@typescript-eslint/restrict-plus-operands': [
```
Comment (311chaos) [inline rules/typescript.js:65]: Relaxing this rule to allow a value that is defined as `any`.

File: .github/workflows/test.yml [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+jobs:
+  tests:
+    runs-on: ubuntu-latest
+  steps:
+  - uses: actions/checkout@v2
+  - uses: actions/setup-node@v2
+    with:
+      node-version: '14'
+      cache: 'npm'
+  - run: npm install
```
Comment (yegorkay) [inline .github/workflows/test.yml:12]: You could cache the npm install like Cypress does here with this action

https://github.com/jane/cypress/blob/main/.github/workflows/check.yml#L18

Comment (311chaos) [reply .github/workflows/test.yml:12]: I wasn't certain what this would cache. Reading the docs again maybe its only caching the deps for this build so it could be tested against a matrix of node versions?

Comment (yegorkay) [reply .github/workflows/test.yml]: Here is see more detail about how/why it works from his blog if it's of any help;

https://glebbahmutov.com/blog/do-not-let-npm-cache-snowball/

Comment (311chaos) [reply .github/workflows/test.yml]: Thanks. I remember reading this now. I'll update and avoid the potential snowball. ;)

**Comment by yegorkay** [comment]
Also, I see codeclimate as a check, and in commerce we have coveralls. Do they not both do the same thing? Can we remove one of these checking services? Just curious.

**Review by yegorkay — APPROVED** [review]

**Comment by 311chaos** [comment]
> 



> Also, I see codeclimate as a check, and in commerce we have coveralls. Do they not both do the same thing? Can we remove one of these checking services? Just curious.

I am not sure I have a good answer on this right now. This repo is open source, so code climate is free here. We used to pay for Coveralls but their service got really poor to where the coverage would not report back for days sometimes, so we cancelled our payments with them. I have thought about removing coveralls from com, but havent made any decision there. If you have suggestions let me know.

**Comment by yegorkay** [comment]
> I have thought about removing coveralls from com, but havent made any decision there. If you have suggestions let me know.

Code quality analysis tooling is a can of worms outside the scope of this PR, so for the sake of this PR, let's keep it for now.

---

Diving deeper:

From my experience, code quality tooling is usually a infra/ops/platform team concern, as well eng SLT. Ideally, an organization should align on a toolkit that works across all services to monitor the health of said services. Here, we picked codeclimate, in commerce we picked coveralls, and I am sure there are several other repos using different analysis tooling. Since I am not on any team relevant to that side of engineering, I am hesitant to call out a solution for this that can work across Jane. 

Even though I am hesitant to give out a solution, I will share that in my previous experience, we have used [codescene](https://codescene.com/), and later migrated to [sonarqube](https://www.sonarqube.org/) for code analysis. These tools were added as a GitHub mandatory check to all the repos we had. If code health decreased a significant amount on a PR, the PR would not pass. These tools are a great investment for an engineering team to align on as it helps everyone quantity the health of the product and gives teams something to improve on quarter to quarter or year to year. I think this is something worthwhile to bring up to the engineering team as a whole, because we all should be using the same tool to monitor the health of the whole company. If we are fragmented, we will never align truly on if our product is in a healthy state regarding the code we push.

**tl;dr** We need Jane engineering to align on a code analysis tool we can use across all repos and services. I do not know what the tool is, but I have some suggestions above. We cannot be fragmented on the code health of the products we own, and aligning on it will help our engineering culture move forward on quality.
